server/src/main/java/org/tctalent/server/service/db/impl/CandidateServiceImpl.java

```diff
@@ -405,2392 +405,2410 @@
+          final List<CandidateExam> candidateExams = candidate.getCandidateExams();
+          if (candidateExams != null) {
+            examO = candidateExams.stream().filter(
+                e -> e.getExam().equals(examType)).findFirst();
+          }
+
+          //Get or create the CandidateExam object
+          CandidateExam exam;
+          if (examO.isPresent()) {
+            exam = examO.get();
+          } else {
+            //Create new exam
+            exam = new CandidateExam();
+            exam.setCandidate(candidate);
+            exam.setExam(examType);
+            exam.setYear((long) OffsetDateTime.now().getYear());
+          }
+
+          //Update existing exam result
+          exam.setScore(answer);
+
+          //Save the exam
+          candidateExamRepository.save(exam);
         } else {
-            throw new InvalidRequestException("Task is not a QuestionTask: " + task.getName());
+          Object answerValue = convertAnswerToCorrectType(answerField, answer);
+          try {
+            PropertyUtils.setProperty(candidate, answerField, answerValue);
+          } catch (IllegalAccessException e) {
+            throw new InvalidRequestException("Unable to access '" + answerField
+                + "' field of candidate");
+          } catch (InvocationTargetException e) {
+            throw new InvalidRequestException("Error while accessing '" + answerField
+                + "' field of candidate");
+          } catch (NoSuchMethodException e) {
+            throw new InvalidRequestException(
+                "Candidate field does not exist: '" + answerField
+                    + "'");
+          }
+
+          save(candidate, true);
         }
-
-        return answer;
+      }
+    } else {
+      throw new InvalidRequestException("Task is not a QuestionTask: " + task.getName());
+    }
+
+    //Fetch answer back to capture any reformatting (eg answers stored as Enums, return their
+    //display name) in the transient answer attached to the task assignment.
+    answer = fetchCandidateTaskAnswer(questionTaskAssignment);
+
+    questionTaskAssignment.setAnswer(answer);
+  }
+
+  /**
+   * Converts the given String answer to the type expected by the given Candidate field. This may be a String, in which case the given
+   * String answer is returned. However, if it is an enum, it will be converted to the matching enum value.
+   *
+   * @param answerField Field of Candidate object
+   * @param answer      String answer to be converted if needed
+   * @return Object containing type version of String answer
+   * @throws InvalidRequestException If we can't find info on the given field on the Candidate class.
+   */
+  private Object convertAnswerToCorrectType(String answerField, String answer)
+      throws InvalidRequestException {
+    PropertyDescriptor pd;
+    try {
+      pd = BeanHelper.getPropertyDescriptor(Candidate.class, answerField);
+    } catch (IntrospectionException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While re-throwing the caught exception with modifications , information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 1302, 1308, 1314.